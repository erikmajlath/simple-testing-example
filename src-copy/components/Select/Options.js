import React from 'react'

export default ({ options, value, onChange }) =>
  <ul>
    {options.map(item => {
      const selected = value.includes(item.value)
      return (
        <li key={item.value} onClick={() => onChange(item, selected)}>
          {item.label}
          <span>
            {selected && ' - SELECTED!'}
          </span>
        </li>
      )
    })}
  </ul>

// Constants
export const SF = { value: 'sf', label: 'San Francisco' }
export const LA = { value: 'la', label: 'Los Angeles' }
export const NY = { value: 'ny', label: 'New York City' }

export const OPTIONS = [SF, LA, NY]
