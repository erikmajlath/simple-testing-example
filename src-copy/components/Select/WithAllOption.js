import React from 'react'

const EMPTY_ARRAY = []
export const ALL_OPTION = { value: 'ALL_OPTION', label: 'All' }

export default WrappedComponent => props => {
  const newProps = {}

  // Add all option to the front
  newProps.options = [ALL_OPTION].concat(props.options)

  // If 'all' is passed as value select all options
  if (props.value === 'all') {
    newProps.value = newProps.options.map(item => item.value)
  }

  // Modify onChange function
  // TODO: move this function outside the component
  newProps.onChange = (value, additionalInfo) => {
    const { option, isSelected } = additionalInfo
    // First we are dealing with all option
    if (option === ALL_OPTION) {
      if (!isSelected) {
        // It has been selected
        return props.onChange('all', additionalInfo)
      } else {
        // It has been deselected
        return props.onChange(EMPTY_ARRAY, additionalInfo)
      }
    }

    // All are selected and we are deselecting
    // Filter away deselected option and select all other
    if (isSelected && props.value === 'all') {
      return props.onChange(
        props.options.filter(item => item !== option).map(item => item.value),
        additionalInfo
      )
    }

    // If all options other then all has been selected, select all
    if (!isSelected && props.options.length === props.value.length + 1) {
      return props.onChange('all', additionalInfo)
    }

    // None of the special cases happened so just pass selected values
    props.onChange(value, additionalInfo)
  }

  return <WrappedComponent {...props} {...newProps} />
}
