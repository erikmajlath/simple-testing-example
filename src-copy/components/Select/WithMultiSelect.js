import React from 'react'

export default WrappedComponent => props =>
  <WrappedComponent {...props} onChange={newOnChange.bind(null, props)} />

const newOnChange = (props, option, isSelected) => {
  const additionalInfo = { option, isSelected }
  let newValue

  if (isSelected) {
    newValue = props.value.filter(value => value !== option.value)
  } else {
    newValue = props.value.concat([option.value])
  }

  props.onChange(newValue, additionalInfo)
}
