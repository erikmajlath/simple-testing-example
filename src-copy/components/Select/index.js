import { compose } from 'redux'

import Options from './Options'
import WithMultiSelect from './WithMultiSelect'
import WithAllOption from './WithAllOption'
export * from './Options'

export default compose(WithAllOption, WithMultiSelect)(Options)
