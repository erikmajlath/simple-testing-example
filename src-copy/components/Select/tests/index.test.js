import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import Select, { OPTIONS, SF, LA, NY } from '../index'
import { ALL_OPTION } from '../WithAllOption'

describe('Select', () => {
  const onChange = jest.fn()
  const component = (
    <Select options={OPTIONS} value={['sf']} onChange={onChange} />
  )

  it('should match snapshot', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('clicks', () => {
    let wrapper

    beforeEach(() => {
      wrapper = mount(component)
      onChange.mockReset()
    })

    it('should call onChange on selected', () => {
      wrapper.find('li').first().simulate('click')

      expect(onChange).toBeCalledWith('all', {
        option: ALL_OPTION,
        isSelected: false,
      })
    })

    it('should call onChange on not selected', () => {
      wrapper.find('li').at(2).simulate('click')

      expect(onChange).toBeCalledWith(['sf', 'la'], {
        option: LA,
        isSelected: false,
      })
    })
  })
})
