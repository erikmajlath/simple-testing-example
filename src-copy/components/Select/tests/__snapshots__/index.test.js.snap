// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should match snapshot 1`] = `
<ul>
  <li
    onClick={[Function]}
  >
    All
  </li>
  <li
    onClick={[Function]}
  >
    San Francisco
     - SELECTED!
  </li>
  <li
    onClick={[Function]}
  >
    Los Angeles
  </li>
  <li
    onClick={[Function]}
  >
    New York City
  </li>
</ul>
`;
